s3fs/s3fs.go

```diff
@@ -0,0 +1,574 @@
+		// to filter the results.
+		// If the contents of 'test/test/' are to be queried, we need to allow
+		// more slashes than we would have to with '/' or 'test/'.
+		var nos int
+		if dirPath == "/" {
+			nos = 1
+		} else {
+			nos = 1 + strings.Count(dirPath, "/")
+		}
+		var isNestedObject bool
+		// Different rules apply to files than to directories.
+		// e.g.:
+		//		key:			slashes			ends with slash
+		// 		test/ 			1 				true
+		// 		test/doc.pdf 	1 				false
+		// These would both only have 1 occurrence of the forward slash.
+		// We want to allow the directory in this case since it is a direct
+		// child of the target directory. We don't want to allow the file.
+		if *c.Size == 0 { // Directory
+			isNestedObject = strings.Count(*c.Key, "/") > nos
+		} else { // File
+			isNestedObject = strings.Count(*c.Key, "/") >= nos
+		}
+		if !recursive && isNestedObject {
+			continue
+		}
+		f := fs.File(*c.Key)
+		fi := s3fs.Stat(f.Name())
+		callback(f, fi)
+	}
+	return nil
+}
+
+// ListDirInfo lists all objects in the given directory and their infos.
+func (s3fs *S3FileSystem) ListDirInfo(dirPath string, callback func(fs.File, fs.FileInfo) error, patterns []string) (err error) {
+	return s3fs.listDirInfo(dirPath, callback, patterns, false)
+}
+
+// ListDirInfoRecursive lists all objects in the given directory recursively.
+func (s3fs *S3FileSystem) ListDirInfoRecursive(dirPath string, callback func(fs.File, fs.FileInfo) error, patterns []string) (err error) {
+	return s3fs.listDirInfo(dirPath, callback, patterns, true)
+}
+
+// ListDirMax lists a max objects in the dirPath.
+func (s3fs *S3FileSystem) ListDirMax(dirPath string, max int, patterns []string) (files []fs.File, err error) {
+	return fs.ListDirMaxImpl(max, func(callback func(fs.File) error) error {
+		return s3fs.ListDirInfo(dirPath, fs.FileCallback(callback).FileInfoCallback, patterns)
+	})
+}
+
+// SetPermissions does not do anything. Setting permissions is not yet
+// implemented.
+// https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/examples-s3-access-permissions.html
+func (s3fs *S3FileSystem) SetPermissions(filePath string, perm fs.Permissions) error {
+	return errors.New("SetPermissions not possible on S3 buckets")
+}
+
+// User is not implemented for S3 file system.
+func (s3fs *S3FileSystem) User(filePath string) string {
+	return ""
+}
+
+// SetUser is not possible for S3 file  system.
+func (s3fs *S3FileSystem) SetUser(filePath string, user string) error {
+	return errors.New("SetUser not possible on S3 buckets")
+}
+
+// Group is not implemented for S3 file system.
+func (s3fs *S3FileSystem) Group(filePath string) string {
+	return ""
+}
+
+// SetGroup is not implemented for S3 file system.
+func (s3fs *S3FileSystem) SetGroup(filePath string, group string) error {
+	return errors.New("SetGroup not possible on S3 buckets")
+}
+
+// Touch creates a file in the S3 bucket (no data).
+func (s3fs *S3FileSystem) Touch(filePath string, perm []fs.Permissions) error {
```

Comment:
Touch should not return an error if the file already exists. See touch unix command!